### PR TITLE
Memoize range_of_any for statements

### PR DESCRIPTION
**Pull request by emjin**
Save the range of statements so that range_of_any is much faster.

Test plan:

make test

PR checklist:
- [x] changelog is up to date

### CODE REVIEW

**Review by mjambon — APPROVED** [review]

File: semgrep-core/src/core/ast/AST_generic.ml [hunk 1]
```diff
@@ -897,7 +897,11 @@ and stmt = {
      and before matching.
   *)
   (* used in semgrep to skip some AST matching *)
-  mutable s_bf : Bloom_filter.t option; [@equal fun _a _b -> true] [@hash.ignore]
+  mutable s_bf : Bloom_filter.t option;
+      [@equal fun _a _b -> true] [@hash.ignore]
+  mutable s_range :
+    (Parse_info.token_location * Parse_info.token_location) option;
+      [@hash.ignore]
```
Comment (mjambon) [inline semgrep-core/src/core/ast/AST_generic.ml:904]: If this field is related to bloom-filter optimization, it should clearly indicate so. A comment + maybe a name containing `bf` would help.

**Review by mjambon — APPROVED** [review]

File: semgrep-core/src/core/ast/AST_generic.ml (hunk 1 above)
Comment (emjin) [reply semgrep-core/src/core/ast/AST_generic.ml:904]: It isn't; `s_range` is just saving the range of the statement

File: semgrep-core/src/core/ast/Visitor_AST.ml [hunk 2]
```diff
@@ -1233,12 +1234,54 @@ let ii_of_any any =
...
-  List.filter Parse_info.is_origintok tokens |> Parse_info.min_max_ii_by_pos
+  List.filter PI.is_origintok tokens |> PI.min_max_ii_by_pos
   [@@profiling]
 
-let range_of_any any =
-  let leftmost_token, rightmost_token = ii_of_any any |> range_of_tokens in
-  ( Parse_info.token_location_of_info leftmost_token,
-    Parse_info.token_location_of_info rightmost_token )
-  [@@profiling]
+let range_of_any any = extract_ranges (fun visitor -> visitor any) [@@profiling]
```
Comment (aryx) [inline semgrep-core/src/core/ast/Visitor_AST.ml:1287]: maybe you could add in the .mli that this function will have some side effects and set the s_range fields for any substmt in any.

**Review by aryx — APPROVED** [review]

**Comment by aryx** [comment]
Do you have a concrete command one can run to see the difference in performance?

**Pull request merged**